modules/infra/control/loopback.c

```diff
@@ -0,0 +1,253 @@
+
+struct iface_info_loopback {
+	int fd;
+	struct event *ev;
+};
+
+static void finalize_fd(struct event *ev, void * /*priv*/) {
+	int fd = event_get_fd(ev);
+	if (fd >= 0)
+		close(fd);
+}
+
+void loopback_tx(struct rte_mbuf *m) {
+	struct mbuf_data *d = mbuf_data(m);
+	struct iface_info_loopback *lo;
+	struct iovec iov[2];
+	struct tun_pi pi;
+	char *data;
+
+	lo = (struct iface_info_loopback *)d->iface->info;
+	if (rte_pktmbuf_linearize(m) == 0) {
+		data = rte_pktmbuf_mtod(m, char *);
+	} else {
+		data = rte_malloc(NULL, rte_pktmbuf_pkt_len(m), 0);
+		if (data == NULL) {
+			LOG(ERR, "rte_malloc failed %s", rte_strerror(rte_errno));
+			goto err;
+		}
+		data = (char *)rte_pktmbuf_read(m, 0, rte_pktmbuf_pkt_len(m), data);
+	}
+	pi.flags = 0;
+	if ((data[0] & 0xf0) == 0x40)
+		pi.proto = RTE_BE16(RTE_ETHER_TYPE_IPV4);
+	else if ((data[0] & 0xf0) == 0x60)
+		pi.proto = RTE_BE16(RTE_ETHER_TYPE_IPV6);
+	else {
+		LOG(ERR, "Bad proto: 0x%x - drop packet", pi.proto);
+		goto err;
+	}
+	// Do not retry even in case of  if EAGAIN || EWOULDBLOCK
+	// If the tun device queue is full, something really bad is
+	// already happening on the management plane side.
+	iov[0].iov_base = &pi;
+	iov[0].iov_len = sizeof(pi);
+	iov[1].iov_base = data;
+	iov[1].iov_len = rte_pktmbuf_pkt_len(m);
+
+	if (writev(lo->fd, iov, ARRAY_DIM(iov)) < 0) {
+		// The user messed up and removed gr-loopX
+		// release resources on our side to try to recover
+		if (errno == EBADFD) {
+			iface_destroy(d->iface->id);
+		}
+		LOG(ERR, "write to tun device failed %s", strerror(errno));
+	}
+
+err:
+	if (!rte_pktmbuf_is_contiguous(m))
+		rte_free(data);
+	rte_pktmbuf_free(m);
+}
+
+static void iface_loopback_poll(evutil_socket_t, short reason, void *ev_iface) {
+	struct eth_input_mbuf_data *e;
+	struct iface_info_loopback *lo;
+	struct iface *iface = ev_iface;
+	struct rte_mbuf *mbuf;
+	size_t len;
+	char *data;
+
+	lo = (struct iface_info_loopback *)iface->info;
+
+	if (reason == EV_CLOSED) {
+		// The user messed up and removed gr-loopX
+		LOG(ERR, "tun device %s deleted", iface->name);
+		iface_destroy(iface->id);
+		return;
+	}
+
+	mbuf = rte_pktmbuf_alloc(loopback_pool);
+	if (!mbuf) {
+		LOG(ERR, "rte_pktmbuf_alloc %s", rte_strerror(rte_errno));
+		goto err;
+	}
+
+	if ((data = rte_pktmbuf_append(mbuf, iface->mtu)) == NULL) {
+		LOG(ERR, "rte_pktmbuf_alloc %s", rte_strerror(rte_errno));
+		goto err;
+	}
+
+	if ((len = read(lo->fd, data, iface->mtu)) <= 0) {
+		if (errno == EAGAIN || errno == EWOULDBLOCK)
+			return;
+		LOG(ERR, "read from tun device %s failed %s", iface->name, strerror(errno));
+		goto err;
+	}
+
+	rte_pktmbuf_trim(mbuf, iface->mtu - len);
+
+	// packet sent from linux tun iface, no need to compute checksum;
+	mbuf->ol_flags = RTE_MBUF_F_RX_IP_CKSUM_GOOD;
+	// Emulate ethernet input, required by ip(6)_input
+	e = eth_input_mbuf_data(mbuf);
+	e->iface = iface;
+	e->eth_dst = ETH_DST_LOCAL;
+
+	post_to_stack(loopback_get_control_id(), mbuf);
```

Comment:
I see no implementation of loopback_get_control_id() in this commit.
Compilation between patches is likely broken.